src/main/java/com/robertx22/mine_and_slash/mechanics/thresholds/SpendThresholdSpec.java

```diff
@@ -0,0 +1,143 @@
+    private final String key;
+
+    // gating/cooldown controls
+    private final Set<String> lockWhileEffectIds;
+    private final int cooldownTicks;
+    private final boolean lockWhileCooldown;         // treat cooldown as a lock
+    private final boolean dropProgressWhileLocked;
+    private final boolean resetProgressOnProc;
+    private final boolean showUi; // whether to render progress HUD for this spec
+
+    // registry ordering (lower runs first)
+    private int priority = 0;
+
+    // Full ctor used by data-driven impl
+    public SpendThresholdSpec(ResourceType resource,
+                              float perLevelFactor,
+                              String key,
+                              Set<String> lockWhileEffectIds,
+                              int cooldownTicks,
+                              boolean lockWhileCooldown,
+                              boolean dropProgressWhileLocked,
+                              boolean resetProgressOnProc,
+                              boolean showUi) {
+        this.resource = resource;
+        this.perLevelFactor = perLevelFactor;
+        this.key = key;
+        this.lockWhileEffectIds = (lockWhileEffectIds == null) ? Collections.emptySet() : Set.copyOf(lockWhileEffectIds);
+        this.cooldownTicks = Math.max(0, cooldownTicks);
+        this.lockWhileCooldown = lockWhileCooldown;
+        this.dropProgressWhileLocked = dropProgressWhileLocked;
+        this.resetProgressOnProc = resetProgressOnProc;
+        this.showUi = showUi;
+    }
+
+    // Backward-compatible ctor (defaults showUi=false)
+    public SpendThresholdSpec(ResourceType resource,
+                              float perLevelFactor,
+                              String key,
+                              Set<String> lockWhileEffectIds,
+                              int cooldownTicks,
+                              boolean lockWhileCooldown,
+                              boolean dropProgressWhileLocked,
+                              boolean resetProgressOnProc) {
+        this(resource, perLevelFactor, key, lockWhileEffectIds, cooldownTicks, lockWhileCooldown, dropProgressWhileLocked, resetProgressOnProc, false);
+    }
+
+    // ===== accessors =====
+    public ResourceType resource()              { return resource; }
+    public String key()                         { return key; }
+    public String keyFor(EntityData unit)       { return key; }
+    public boolean lockWhileCooldown()          { return lockWhileCooldown; }
+    public boolean dropProgressWhileLocked()    { return dropProgressWhileLocked; }
+    public boolean resetOnProc()                { return resetProgressOnProc; }
+    public int cooldownTicks()                  { return cooldownTicks; }
+    public int priority()                       { return priority; }
+    public boolean showUi()                     { return showUi; }
+
+    // fluent config (for code-defined specs)
```

Comment:
These lines from 73 to 111 still needed to be removed but save `withPriority` method